[Build] iOS simulator build intermittently performs a clean build (when it should be incremental)

emw-apple opened the pull request:
#### 3a0f6f625177592a3b5f5458ab39209a923ff70f
<pre>
[Build] iOS simulator build intermittently performs a clean build (when it should be incremental)
<a href="https://rdar.apple.com/127636159">rdar://127636159</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=274184">https://bugs.webkit.org/show_bug.cgi?id=274184</a>

Reviewed by Alexey Proskuryakov.

Incremental builds break when the targeted simulator changes. Xcode sets
build settings like TARGET_DEVICE_MODEL and TARGET_DEVICE_IDENTIFIER
based on destination, so when the selected simulator changes, all our
build script executions are invalidated and the next build rebuilds
nearly everything.

webkitdirs was picking the first eligible simulator as reported by
`simctl list devices`. This was causing incremental rebuilds when
engineers add and remove simulator devices (relatively common during
testing workflows). Additionally, it meant that switching between the
Xcode IDE and command line for building was likely to be non-incremental
-- you&apos;d have to know which simulator was the special one and select it
in the UI.

Fix by teaching webkitdirs to look at previous build requests and find
the UDID of the simulator that was used. If it&apos;s still available and
eligible, build with it. If it isn&apos;t, pick an eligible simulator -- the
next request will re-use it. Instead of picking simulators based on the
order simctl reports them, sort lexicographically like Xcode does, so
that the default matches the first list item in the UI.

* Tools/Scripts/webkitdirs.pm:
(determineXcodeDestination):

Canonical link: <a href="https://commits.webkit.org/278893@main">https://commits.webkit.org/278893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc230839fc74365db9b7be35e6769845aa9c07f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51847 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31159 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4211 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2539 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/53946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28779 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44721 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23307 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1983 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45189 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2075 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56706 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51353 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26968 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1954 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44821 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11332 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29102 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/63672 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12016 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -617,16 +628,46 @@ sub determineXcodeDestination
+        
+        # If we found the previous device, check that the runtime being built has not changed (e.g. due to a
+        # major SDK update). If it has changed, or if no previous device is available, fall back to the first
+        # eligible device in the list (matching legacy behavior).
```
How does Xcode decide which device to build for? It would be best to match it, not legacy behavior.

I'm thinking of the case of building with `make` first, and then with Xcode. It would be best to avoid rebuilding in this case, too.

emw-apple replied:
Xcode doesn't decide, you have to pick a destination. Though the list in the UI appears to be in reverse order of what `xcrun simctl list devices` shows. Perhaps we should default to the _last_ eligible device, so that it's akin to picking the first device in the UI. And perhaps we should print the device name we're using so that power users can pick it manually.

emw-apple replied:
> Though the list in the UI appears to be in reverse order of what `xcrun simctl list devices` shows.

It's actually lexicographic; the patch I'm landing does this too.

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -617,16 +628,46 @@ sub determineXcodeDestination
+        my @devices = iOSSimulatorDevices();
+        my $prevDevice = first { $prevUDID && $_->{UDID} eq $prevUDID } @devices;
+        if ($prevUDID && !$prevDevice) {
+            warn "Simulator with UDID '$prevUDID' not found, falling back to another available simulator. " .
```
I wonder if these should be more noticeable (with extra blank lines perhaps). Or are they already printed early enough that they are noticeable?

emw-apple replied:
They would usually be the first lines printed after running `make`, because they show up before xcodebuild is invoked. But I'm open to adding more formatting / indication that these are WebKit-specific messages and not xcodebuild noise.

emw-apple replied:
Going to hold off on this, I'd rather explore it in a follow-up where we refactor `set-webkit-configuration` to share some more UI code with webkitdirs itself.

> They would usually be the first lines printed after running `make`, because they show up before xcodebuild is invoked.

Correction: it's the first line after the "===== BUILDING <scheme> =====" message. I still think this is high up enough to be noticeable for someone investigating why their build is slow.

aproskuryakov reviewed (approved):
Makes sense, please consider these enhancements.

emw-apple committed:
[Build] iOS simulator build intermittently performs a clean build (when it should be incremental)
rdar://127636159
https://bugs.webkit.org/show_bug.cgi?id=274184

Reviewed by Alexey Proskuryakov.

Incremental builds break when the targeted simulator changes. Xcode sets
build settings like TARGET_DEVICE_MODEL and TARGET_DEVICE_IDENTIFIER
based on destination, so when the selected simulator changes, all our
build script executions are invalidated and the next build rebuilds
nearly everything.

webkitdirs was picking the first eligible simulator as reported by
`simctl list devices`. This was causing incremental rebuilds when
engineers add and remove simulator devices (relatively common during
testing workflows). Additionally, it meant that switching between the
Xcode IDE and command line for building was likely to be non-incremental
-- you'd have to know which simulator was the special one and select it
in the UI.

Fix by teaching webkitdirs to look at previous build requests and find
the UDID of the simulator that was used. If it's still available and
eligible, build with it. If it isn't, pick an eligible simulator -- the
next request will re-use it. Instead of picking simulators based on the
order simctl reports them, sort lexicographically like Xcode does, so
that the default matches the first list item in the UI.

* Tools/Scripts/webkitdirs.pm:
(determineXcodeDestination):

Canonical link: https://commits.webkit.org/278893@main
